### ISSUE TITLE
internally_dereference_refs generates wrong result when discriminators are used in spec

### ISSUE THREAD

**Issue by gstarnberger** [Yelp/bravado-core#210]
internally_dereference_refs appears to not properly deserialize responses when discriminators are used in the spec and when use_models is enabled. In particular, if the model contains a type and a subtype none of the attributes of the subtype gets deserialized.

Here's an example spec: https://gist.github.com/gstarnberger/4dd7d76b94d7afec87f8a6a141dfd17e

If you call an endpoint that returns `{"score":12345,"type":"subtypeOfResponse"}` as response that response gets deserialized as:

- `subtypeOfResponse(score=12345, type=u'subtypeOfResponse')` (with internally_dereference_refs disabled)
- `subtypeOfResponse()` (with internally_dereference_refs enabled)

Directly accessing the missing attributes still yields a value, but that value doesn't use any models (although use_models is set to True). For example, if score would have been defined as another object type then subtypeOfResponse.score would have returned a dictionary.

Here's the code that I used to reproduce the issue:

Server: https://gist.github.com/gstarnberger/359669829881ca5fa188f5a480820512 (basically just returns the same hardcoded response for any call)

Client: https://gist.github.com/gstarnberger/3701dba50aba719ffc0bf28454485f31

Output of the client when running the code is:

```
False subtypeOfResponse(score=12345, type=u'subtypeOfResponse')
True subtypeOfResponse()
```

**Comment by macisamuele** [comment]
@gstarnberger I've started to dig into the issue that you're describing.
I tried to formalize your scenario into a single test so we can add it to the _normal_ test suite in order to prevent future regression.
https://gist.github.com/macisamuele/4fdb2a320a95f6aa24ecef9f6d16248d

At the moment I'm getting validation error while using fully dereferenced specs (note you have disabled response validation).
It means that somehow the validation logic has an issue or we should not use fully dereferenced specs for specs that take advantage of discriminators. I will look deeper into it as soon as I have additional time.

Thanks for reporting the issue.
